src/ripple/rpc/handlers/DownloadShard.cpp

```diff
@@ -129,19 +129,24 @@ doDownloadShard(RPC::JsonContext& context)
         }
     }
 
-    RPC::ShardArchiveHandler::pointer handler;
+    RPC::ShardArchiveHandler* handler = nullptr;
 
     try
     {
-        handler = RPC::ShardArchiveHandler::hasInstance()
-            ? RPC::ShardArchiveHandler::getInstance()
-            : RPC::ShardArchiveHandler::getInstance(
-                  context.app, context.app.getJobQueue());
+        handler = context.app.getShardArchiveHandler();
 
         if (!handler)
             return RPC::make_error(
                 rpcINTERNAL, "Failed to create ShardArchiveHandler.");
 
+        if (handler->processingDownloads())
+        {
+            std::string s{"Download already in progress. Shard"};
+            if (!std::all_of(preShards.begin(), preShards.end(), ::isdigit))
+                s += "s";
+            return RPC::makeObjectValue(s + " " + preShards);
+        }
+
         if (!handler->init())
```

Comment:
I added a check in `init`. I think it's still good to have the explicit call to `isProcessing` to abort the rpc command early if possible.

Reply:
I don't think adding the check in `init` is sufficient, since we're not setting `process_` there (and can't since `add` requires it to be unset). In addition to the check in `init`, I will also add a lock when we set `downloader_` and `sqliteDB_` and return false if either is already set. This makes sure that initialization won't have a race condition.